snapshot/git/gitdb/local_data.go

```diff
@@ -108,13 +109,76 @@ func (db *DB) ingestDir(dir string) (snapshot.ID, error) {
 	return v.ID(), nil
 }
 
+const tempRef = "refs/heads/scoot/__temp_for_writing"
+
+func (db *DB) ingestGitCommit(ingestRepo *repo.Repository, commitish string) (snapshot.ID, error) {
+	sha, err := ingestRepo.RunSha("rev-parse", "--verify", fmt.Sprintf("%s^{commit}", commitish))
+	if err != nil {
+		return "", fmt.Errorf("not a valid commit: %s, %v", commitish, err)
+	}
+
+	// Strategy:
+	// trying to move a commit from ingest to data
+	// delete the ref in the data.
+	// set the ref in the ingest.
+	// push from ingest to data.
+	// delete in both repos.
+	// TODO(dbentley): we could check if sha exists in our repo before ingesting
+
+	if _, err := db.dataRepo.Run("update-ref", "-d", tempRef); err != nil {
+		return "", err
+	}
+
+	if _, err := ingestRepo.Run("update-ref", tempRef, sha); err != nil {
+		return "", err
+	}
+
+	if _, err := ingestRepo.Run("push", "-f", db.dataRepo.Dir(), tempRef); err != nil {
+		return "", err
+	}
+
+	if _, err := ingestRepo.Run("update-ref", "-d", tempRef); err != nil {
+		return "", err
+	}
+
+	if _, err := db.dataRepo.Run("update-ref", "-d", tempRef); err != nil {
+		return "", err
+	}
+
+	l := &localValue{sha: sha, kind: kindSnapshotWithHistory}
+	return l.ID(), nil
+}
+
+// checkout creates a checkout of id.
 func (db *DB) checkout(id snapshot.ID) (path string, err error) {
-	// Checkout creates a new dir with a new index and checks out exactly that tree.
+
+	defer func() {
+		if path != db.dataRepo.Dir() {
```

Comment:
Commented:

		// If we're our repo dir, we need to keep the work tree locked.
		// Otherwise, we can unlock it.